src/hydra-queue-runner/hydra-queue-runner.cc

```diff
@@ -449,74 +463,131 @@ bool State::checkCachedFailure(Step::ptr step, Connection & conn)
-                        break;
-                };
-                execvp("hydra-notify", (char * *) stringsToCharPtrs(argv).data()); // FIXME: remove cast
-                throw SysError("cannot start hydra-notify");
-            });
-
-            int res = pid.wait();
-
-            if (!statusOk(res))
-                throw Error("notification about build %d failed: %s", item.id, statusToString(res));
-
-            auto now2 = std::chrono::steady_clock::now();
-
-            if (item.type == NotificationItem::Type::BuildFinished) {
-                auto conn(dbPool.get());
-                pqxx::work txn(*conn);
-                txn.parameterized
-                    ("update Builds set notificationPendingSince = null where id = $1")
-                    (item.id)
-                    .exec();
-                txn.commit();
+        FILE* toNotify = fdopen(dup(to.writeSide.get()), "w");
+        FILE* fromNotify = fdopen(dup(from.readSide.get()), "r");
+
+        new_pid = startProcess([&]() {
+
+            if (dup2(to.readSide.get(), STDIN_FILENO) == -1) {
+              throw SysError("cannot dup input pipe to stdin");
+            }
+            if (dup2(from.writeSide.get(), STDOUT_FILENO) == -1) {
+              throw SysError("cannot dup output pipe to stdout");
             }
 
-            nrNotificationTimeMs += std::chrono::duration_cast<std::chrono::milliseconds>(now2 - now1).count();
-            nrNotificationsDone++;
+            execlp("hydra-notify", "hydra-notify", NULL);
 
-        } catch (std::exception & e) {
-            nrNotificationsFailed++;
-            printMsg(lvlError, format("notification sender: %1%") % e.what());
-            sleep(5);
-        }
+            throw SysError("cannot start hydra-notify");
+        });
+
+        close(to.readSide.release());
+        close(from.writeSide.release());
+
+        hydra_notify = std::optional<hydra_notify_state> {{
+          .pid = new_pid,
+          .to_notify = std::move(to),
+          .from_notify = std::move(from),
+          .to_notify_stream = toNotify,
+          .from_notify_stream = fromNotify,
+          .buf_notify = NULL
+        }};
+      }
+
+      NotificationItem item;
+      {
+          auto notificationSenderQueue_(notificationSenderQueue.lock());
+          while (notificationSenderQueue_->empty())
+              notificationSenderQueue_.wait(notificationSenderWakeup);
+          item = notificationSenderQueue_->front();
+          notificationSenderQueue_->pop();
+      }
+
+
+      MaintainCount<counter> mc(nrNotificationsInProgress);
+
+      printMsg(lvlChatty, format("sending notification about build %1%") % item.id);
+
+      auto now1 = std::chrono::steady_clock::now();
+
+      std::string payload;
+
+      switch (item.type) {
+        case NotificationItem::Type::BuildStarted:
+          payload = "build-started"s + " "s + std::to_string(item.id);
```

Comment:
The more idiomatic way to do this would be to use either `osstream` (from `sstream` header) or to do the interpolation in the `fprintf` you're already doing soon.